### ISSUE TITLE
MDC/GDC/MDLC improvements

### ISSUE THREAD

**Issue by 304NotModified** [NLog/NLog#886]
- add indexer syntax, e.g. 

```c#
MappedDiagnosticsContext["a"] = "b";
var x= MappedDiagnosticsContext["a"];
```
- One common interface for MDC/GDC/MDLC: proposal: `NLogContext`
- [ ] Remove " Mostly for compatibility with log4net." comments from code 
- [ ] Remove " Mostly for compatibility with log4net." comments from wiki 
- Possible: implement `IDictionary<string, object>` ?

note: NDC has different methods

**Comment by 304NotModified** [comment]
related: #873

**Comment by SreenathSanthanam** [comment]
I am currently working on this.

**Comment by SreenathSanthanam** [comment]
obsolete warnings for what methods needs to be added?

Also interface segregation is not possible with backward compatibility.

**Comment by 304NotModified** [comment]
> Also interface segregation is not possible with backward compatibility.

what do you mean? I was thinking about 

c#
```
interface IContext
{
 Dictionary<string, object>.KeyCollection Keys //I think we like to use List or Hashset?
 object this[string key]
 void Clear()
 bool Contains(string item)
 void Remove(string item) //or bool instead of void?
}

**Comment by SreenathSanthanam** [comment]
You are correct. I did the same. To do like this you will have to remove the static methods which would cause backward im-compatible.

**Comment by 304NotModified** [comment]
Good point. Maybe we need:

Different method names or a new class beside this one.

**Comment by SreenathSanthanam** [comment]
Something in the line of GlobalContext? and similar names for other Context Classes as well.

**Comment by 304NotModified** [comment]
Sounds good. The question is if we should rename mdc just to threadcontext 

And maybe even skip mdc code and replace with MDLC code

What do you think?

**Comment by SreenathSanthanam** [comment]
Yes that sounds about right.

So the consolidated list will have 

GlobalContext
MDC and MDLC merged into one. 

The other question
Do we remove the older MDC, GDC, Classes?

**Comment by 304NotModified** [comment]
We can only remove things in 5.0,  according to semver.

**Comment by Page-Not-Found** [comment]
I am facing a peculiar unit test failure. Which passes in my local PC, but fails in appveyor 4.5, the same passes for 4.0

https://ci.appveyor.com/project/Page-Not-Found/nlog/build/4.0.11

@304NotModified do you have any idea?

**Comment by 304NotModified** [comment]
I will check

**Comment by 304NotModified** [comment]
Well it's using threads, so locks etc are important. Is something changed in the MappedDiagnosticsLogicalContext?

**Comment by Page-Not-Found** [comment]
Yes, I re factored to point to new Logical Thread Context Class

**Comment by Page-Not-Found** [comment]
TC "given_multiple_threads_running_asynchronously_when_setting_and_getting_values_should_return_thread_specific_values" in MappedDiagnosticsLogicalContextTests is failing for some reason in the NLog branch itself.

**Comment by 304NotModified** [comment]
I never saw failing that test the last year?

**Comment by Page-Not-Found** [comment]
Can you check the TC in you local machine and let me know if it is failing or passing. I just now took the NLog Master latest and ran the TC in my machine.

**Comment by 304NotModified** [comment]
It succcees when running all the tests of `MappedDiagnosticsLogicalContextTests`. If I run it separate, it indeed fails, but not the first time.

Ran the tests around 20 times.

**Comment by Page-Not-Found** [comment]
Yes, what should be done now? Do we skip the TC or skip the refactoring of the class itself?

**Comment by 304NotModified** [comment]
maybe it is https://github.com/NLog/NLog/issues/934 ?

**Comment by 304NotModified** [comment]
I propose looking at https://github.com/NLog/NLog/issues/934 first (new branch). If we can easily fix it, then that have the preference. If not, well, we see then :)

**Comment by Page-Not-Found** [comment]
I have read #934 I will try to get this suggestion implemented for the refactored class and run the test again.

**Comment by 304NotModified** [comment]
:+1:

**Comment by Page-Not-Found** [comment]
The design in the blog referred in #934 looks much better. I will see if that can be implemented.

**Comment by 304NotModified** [comment]
Thanks for the up!

**Comment by Page-Not-Found** [comment]
I was looking for Dictionary which is immutable for MDLC, it is available only in .Net4.5 and above. To pursue this idea I was looking for Custom implemented ReadOnlyDict like ReadOnlyCollection and found the this [article](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=29)

I wasn't sure if we can use this code from the above blog since there was mention about License terms, so I wanted to confirm with you guys before I use the code in NLog.

**Comment by 304NotModified** [comment]
License looks good,  but I not sure if we need it. Also (note: didn't looked at depth of it), most code from blogs isn't well tested (compared to github repository)

**Comment by Page-Not-Found** [comment]
I was just going to use it as an example for creating immutable objects.

**Comment by 304NotModified** [comment]
:+1:

**Comment by Page-Not-Found** [comment]
DISCUSSION: What is the plan for phasing out 4.0 support?

**Comment by Page-Not-Found** [comment]
@304NotModified @BrutalCode I would like to discuss with you guys about rewriting GDLC. I am not able to come up with a Good Solution for refactoring. 

* If we take the path suggested in #934 then we wouldn't be able to maintain the same Dictionary structure as we currently have in place. 
* If we want to maintain the dict structure the we would need to Use a ReadOnlyCollection which is supported only 4.5 and above, in this case we wouldn't be able to support 4.0 for the GDLC.

**Comment by BrutalCode** [comment]
[ReadOnlyCollection<T>] (https://msdn.microsoft.com/en-us/library/ms132474(v=vs.100).aspx) is supported 2.0 and above. 

`ReadOnlyCollection` makes an array read-only. But we have `Get` and `Set` methods. 
I am not sure if I understand correctly.

**Comment by Page-Not-Found** [comment]
Sorry for misstating the words. What I meant is [Immutable Collection](https://msdn.microsoft.com/en-us/library/dn467185(v=vs.111).aspx)

**Comment by BrutalCode** [comment]
OK. I get it now. It has a dependency  [System.Collections.Immutable] (https://www.nuget.org/packages/System.Collections.Immutable/)
System.Collections.Immutable's dependenices are in dotnet 5.0

**Comment by Page-Not-Found** [comment]
Now we have to decide whether to re-factor the class without the support for 4.0 and have the support for 4.5 above.

**Comment by BrutalCode** [comment]
Yes. So we need to think about that its usage. Probably most of people are using GDC/MDC in 4.0 and below, and we should refactor by taking into consideration.

**Comment by Page-Not-Found** [comment]
The item in question now is MDLC.

**Comment by BrutalCode** [comment]
Sorry :) MDLC was supported 4.5. [I added 4.0 at aug.] (https://github.com/NLog/NLog/commit/d231b2dea84fc1f5de6408b59ed25046d1d4b169)

Maybe we can remove 4.0 supporting?

**Comment by Page-Not-Found** [comment]
Do others have any opinion?

**Comment by 304NotModified** [comment]
I prefer keeping support for .Net 4. Otherwise it is a breaking change.

**Comment by Page-Not-Found** [comment]
So we remove the GDLC refactoring for now and see at a later point for refactoring?

**Comment by 304NotModified** [comment]
Yes, that's a good approach.

**Comment by Page-Not-Found** [comment]
Thanks. I will revert the changes i Did with respect to GDLC and send for a PR.

**Comment by 304NotModified** [comment]
Is it possible to extract the unit tests (with current code base) to a new PR? Then I will include those first.

**Comment by Page-Not-Found** [comment]
Sorry I am not sure, In understand you clearly. Are you asking me to send the PR without the code revert?

**Comment by 304NotModified** [comment]
No reverts needed.

1. Create a new branch, (from right history)
2. copy the unit tests that will work in 4.2 
3. create second PR. 

:)

**Comment by Page-Not-Found** [comment]
Ah. Okie. I already created a second PR. Can you take a look at that quickly and let me know if I stil need to the do the above?

**Comment by dnlgmzddr** [comment]
I think one of the best features on NLOG is the wide support of frameworks, so I think we should keep supportihng 4.0.

**Comment by 304NotModified** [comment]
> I think one of the best features on NLOG is the wide support of frameworks, so I think we should keep supportihng 4.0.

@dnlgmzddr I'm not sure we you are responding to.

**Comment by 304NotModified** [comment]
https://github.com/NLog/NLog/pull/1060 has been merged to branch "Overhaul-contexts".

**Comment by 304NotModified** [comment]
There is not much development here isn't?

**Comment by kidchenko** [comment]
This issue still planned for 5.0 release?

**Comment by 304NotModified** [comment]
I really like this for 5.0, but there isn't started on the development yet.

**Comment by snakefoot** [comment]
Could be neat if GDC / Variables lived on the LogFactory, so one could have multiple isolated instances side-by-side. See also my comments here: #1429

**Comment by 304NotModified** [comment]
Would be nice if there was a ${context} render which looks in all contexes ;)

**Comment by snakefoot** [comment]
.NET 10 (C# ver. 14) will add support for extension-methods to static-classes:

- https://devblogs.microsoft.com/dotnet/csharp-exploring-extension-members/

Not sure I like the syntax, but probably because I'm old.
